src/MultiTokenPermit.sol

```diff
@@ -0,0 +1,237 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.0;
+
+import { IERC1155 } from "@openzeppelin/contracts/token/ERC1155/IERC1155.sol";
+import { IERC721 } from "@openzeppelin/contracts/token/ERC721/IERC721.sol";
+
+import { IMultiTokenPermit } from "./interfaces/IMultiTokenPermit.sol";
+
+import { PermitBase } from "./PermitBase.sol";
+
+/**
+ * @title MultiTokenPermit
+ * @notice Multi-token support (ERC20, ERC721, ERC1155) for the Permit3 system
+ * @dev Extends PermitBase with NFT and semi-fungible token functionality
+ */
+abstract contract MultiTokenPermit is PermitBase, IMultiTokenPermit {
+    /**
+     * @notice Query multi-token allowance for a specific token ID
+     * @param owner Token owner
+     * @param token Token contract address
+     * @param spender Approved spender
+     * @param tokenId Token ID (0 for ERC20, type(uint256).max for collection-wide wildcard)
+     * @return amount Approved amount (max uint160 for unlimited)
+     * @return expiration Timestamp when approval expires (0 for no expiration)
+     * @return timestamp Timestamp when approval was set
+     */
+    function allowance(
+        address owner,
+        address token,
+        address spender,
+        uint256 tokenId
+    ) external view override returns (uint160 amount, uint48 expiration, uint48 timestamp) {
+        if (tokenId == 0 || tokenId == type(uint256).max) {
+            // For ERC20 tokens or collection-wide wildcard, use the standard allowance
```

Comment:
What do you mean by "collection-wide wildcard"?

Reply:
Means can transfer my NFT token address with any ID I have. Versus allowance to only transfer that one NFT with ID#420